Implementation of part sourcer + tests, added size field to part

ihc opened the pull request.

Commit:
Initial sourcer prototype.

ihc committed:
Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer
Also rewrote the sourcer to save parts in the db.

ihc committed:
Implemented Seagate-specific metadata extraction. Fixed link problem.

ihc committed:
Refactored crawler design (e.g. no more reflection, woot).

ihc committed:
Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer

ihc committed:
Renamed packages, some cleanup, better variable names, etc.

ihc committed:
Added size and description fields to Part.

Modified tests to suit, added size/desc extraction to SeagateCrawler.

ihc committed:
Added unit test for crawler controller.

ihc committed:
Now Java 7 compatible (Stream to Collection), full Controller test.

Also some moving around of crawler dependencies.

ihc committed:
Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer

Conflicts:
	.gitignore
	server/build.gradle
	server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Part.java
	server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/servlet/Servlet.java

ihc committed:
Added size to part metadata, split metadata extraction out into its own class.

ihc committed:
Conflict resolving, merging, renaming, unpleasant things in general.

ihc committed:
Changed various classes for compatibility with the new Part setup (size).

ihc committed:
Teeeessssttts. For sourcer, crawlers, controllers, metadata, size.

ihc committed:
Various small fixes, added manufacturer id getter, better tests ...

Commit:
Changed auto-scan to manual configure

ihc committed:
Some final changes, comments n stuff.

ihc committed:
Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer

ihc committed:
Changed Collection<Part> to List<Part> for compatibility with bulk database saving.

Veedrac commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Size.java:
```diff
@@ -0,0 +1,100 @@
+     */
+    public Size() {
+        dimension1 = 0;
+        dimension2 = 0;
```
`0` seems rather close to a valid value; I'm not sure it's sensible to be using it as a null state. Why aren't you just using `null` instead of empty `Size` objects though?

Alternatively you could make the `getDimension` methods throw when `unit` is `Unit.UNKNOWN`. I guess you'd also need a stronger way to check for having a single valid dimension, but that's not too hard. You could use `Float`s and set them to `null`.

ihc replied:
Because then I have to deal with lots of null checking :]
It has been done in the new version though; but I still initialise an empty Size object for extrusion sizes that couldn't be parsed, because it gives a nicer message on printout.

as2388 commented on server/build.gradle:
```diff
@@ -14,18 +14,24 @@ dependencies {
     compile 'org.mongodb:mongo-java-driver:2.12.2'
 
-    // MongoJack (used for converting Java objects into MongoDB's format)
+    // MongoJack (used for converting Java objects into MongoDB's format
```
I want my bracket back

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/configuration/Configuration.java:
```diff
@@ -15,5 +21,41 @@
 
+    private static final String CRAWL_STORAGE_FOLDER = "crawlerdata/root";
+    private static final int MAX_CRAWL_DEPTH = 5;
+    private static final int MAX_CRAWL_PAGES = -1;
```
Why is this negative?

as2388 replied:
These can be public static final

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Part.java:
```diff
@@ -77,4 +108,14 @@ public boolean equals(Object o) {
             return false;
         }
     }
+
```
The equals function needs to be updated

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Part.java:
```diff
@@ -77,4 +108,14 @@ public boolean equals(Object o) {
     }
+
+    @Override
+    public int hashCode() {
```
Find out if we still need this

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Size.java:
```diff
@@ -0,0 +1,100 @@
+    /**
+     * Constructor for two-dimensional size (e.g. width x height).
+     */
+    public Size(float dim1, float dim2, Unit unit) {
```
Should we be allowed to define dim2 but not dim1?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Size.java:
```diff
@@ -0,0 +1,100 @@
+            return false;
+        }
+
+        Size size = (Size)o;
```
whitespace

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcer.java:
```diff
@@ -0,0 +1,124 @@
+        LoggerFactory.getLogger(PartSourcer.class);
+
+    /**
+     * commandline invocation of the part sourcer.
```
Should you doc what the args should be?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcer.java:
```diff
@@ -0,0 +1,124 @@
+     */
+    public static void main(String[] args) throws Exception {
+
+        String dbName = args.length > 0 ? args[0] : "extrusionDB";
```
change : "extrusionDB" to Configuration.DEFAULT_DATABASE_NAME

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcer.java:
```diff
@@ -0,0 +1,124 @@
+
+        String dir = Configuration.getCrawlStorageFolder();
+        int dp = Configuration.getMaxCrawlDepth();
+        int pg = Configuration.getMaxCrawlPages();
```
I don't like such short variable names

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcer.java:
```diff
@@ -0,0 +1,124 @@
+        } catch (CrawlerException | IllegalArgumentException e) {
+
+            logger.error(e.getLocalizedMessage());
+            System.exit(1);
```
I would rather not have a bad update take down the servlet

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcer.java:
```diff
@@ -0,0 +1,124 @@
+        for (Controller<? extends ExtendedCrawler> c : controllers) {
+
+            try {
+                List<Part> parts = c.start();
```
c.start() returns data? This is weird.

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcer.java:
```diff
@@ -0,0 +1,124 @@
+            } catch (Exception e) {
+
+                logger.error(e.getLocalizedMessage());
+                System.exit(1);
```
yuck

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/Controller.java:
```diff
@@ -0,0 +1,86 @@
+    private T crawler;
+    private CrawlController controller;
+
+    private static final int NUMBER_OF_CRAWLERS = 1;
```
Is this something which should be in the configuration?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/Controller.java:
```diff
@@ -0,0 +1,86 @@
+     */
+    public void stop() {
+        controller.shutdown();
+        controller.waitUntilFinish();
```
Do we have guarantees that it will finish? Do we need to think about force termination?

Veedrac replied:
Is this a problem? It's a standalone application, so we can just kill it manually assuming that the database is handled safely.

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/ExtendedCrawler.java:
```diff
@@ -0,0 +1,31 @@
+    /**
+     * @return  URL seeds for this crawler.
+     */
+    public abstract String[] getSeeds();
```
Since this is an abstract class, why not define SEEDS and MANUFACTURER_ID in this class?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java:
```diff
@@ -0,0 +1,201 @@
+
+    /* We are not interested in entering pages with these file endings */
+    private final static Pattern FILTERS =
+        Pattern.compile(".*(\\.(css|js|gif|jpe?g|png|mp3|mp3|zip|gz|pdf))$");
```
Do we really want to process all these endings?

Veedrac replied:
This is a blacklist; shouldn't you be preferring a whitelist?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java:
```diff
@@ -0,0 +1,201 @@
+    /* Extracts metadata from the page. Add found parts to the internal list
+     * of parts.
+     */
+    private void extractData(MetadataParser hp) {
```
hp?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java:
```diff
@@ -0,0 +1,201 @@
+
+        String url = page.getWebURL().getURL();
+
+        if (page.getParseData() instanceof HtmlParseData) {
```
What else can it be an instanceof? And is it ok to ignore those cases?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java:
```diff
@@ -0,0 +1,201 @@
+
+            // get the image of the product
+            String image = hp.selectSingleAttr(productNode, "img[src]", "abs:src");
+            if (!image.isEmpty()) {
```
What does it mean for the image to be empty? Can we keep processing this part if it is?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java:
```diff
@@ -0,0 +1,201 @@
+     * images_catalog/XYZ pdf (1).pdf
+     * where XYZ is the product id. Some of their own (!) urls on their
+     * website are wrong and don't include the " pdf (1)" at the
+     * end ... This is a temporary solution.
```
"temporary". sure...

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/SeagateCrawler.java:
```diff
@@ -0,0 +1,201 @@
+        for (Pattern p : METADATA_FILTERS) {
+
+            Matcher m = p.matcher(description);
+            if (m.find()) {
```
if (m.find()), not while(m.find())?

ihc replied:
m is dependent on p. It tries a pattern, if that pattern fails, it goes on to the next, and so on. When it finds a match, it breaks the loop, so there's no need for while.

I used longer names in the new version so it is more noticeable.

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/metadata/MetadataParser.java:
```diff
@@ -0,0 +1,172 @@
+     *                  given selector. If none found, the empty string.
+     *
+     * E.g. element contains <div data="hey">This is a value</div> and select
+     * is "div" will return "hey".
```
Can we have an abstract description as well as documentation-by-example?

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/metadata/MetadataParser.java:
```diff
@@ -0,0 +1,172 @@
+     * @return          The attribute value of the first found result for the
+     *                  given selector. If none found, the empty string.
+     *
+     * E.g. element contains <div data="hey">This is a value</div> and select
```
I think this should be part of the @return description

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/metadata/MetadataParser.java:
```diff
@@ -0,0 +1,172 @@
+     * @return  Int parsed from the string.
+     * @throws MetadataParserException  if it was not possible to parse the string.
+     */
+    public static int stringToInt(String s) throws MetadataParserException {
```
Why are you calling stringToInt() instead of just using Integer.parseInt()?

ihc replied:
encapsulaaation. More informative error message + to know the source of the error if it happens.

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/metadata/MetadataParser.java:
```diff
@@ -0,0 +1,172 @@
+                return Unit.MM;
+
+            default:
+                return Unit.UNKNOWN;
```
Should we log what wasn't matched? (Then we check if the above are missing any cases)

as2388 commented on server/src/test/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/SizeTester.java:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    /**
+     * Tests that sizes initialised to a negative value are just set to 0.
```
Why not throw IllegalArgumentException?

as2388 commented on server/src/test/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/PartSourcerTester.java:
```diff
@@ -0,0 +1,117 @@
+     * NOTE: uses mockito!
+     */
+    @Test
+    @SuppressWarnings("unchecked")
```
Add a justifying comment

as2388 commented on server/src/test/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/SourcerIntegrationTester.java:
```diff
@@ -0,0 +1,143 @@
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({ Configuration.class} )
+@PowerMockIgnore( {"javax.management.*"}) // this prevents some weird exception ...
```
Maybe this comment could be better

as2388 commented on server/src/test/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/SourcerIntegrationTester.java:
```diff
@@ -0,0 +1,143 @@
+     * (unit) tests.
+     */
+    @Test
+    @SuppressWarnings("unchecked")
```
why

as2388 commented on server/src/test/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/SourcerIntegrationTester.java:
```diff
@@ -0,0 +1,143 @@
+     */
+    @Test
+    @SuppressWarnings("unchecked")
+    public void testPartSourcer() throws Exception {
```
explain throws Exception

as2388 commented:
Your JavaDoc doesn't compile

as2388 commented on server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/sourcer/crawlers/CrawlControllerFactory.java:
```diff
@@ -0,0 +1,41 @@
+import edu.uci.ics.crawler4j.robotstxt.RobotstxtConfig;
+import edu.uci.ics.crawler4j.robotstxt.RobotstxtServer;
+
+/* it's a factory!*/
```
How about some additional "sensible" javadoc

as2388 commented:
For 100% coverage you need tests for the default constructors

ihc committed:
Added lots of comments, fixes, to address issues raised in pull request comments.

Moved manufacturer information out to its own class, changed a bunch of stuff in Size.

ihc committed:
fine ... you may have your bracket ... <.<

ihc committed:
Some extra small comments and stuff.

ihc committed:
Changed parts to be a set to avoid duplicates (for database saving).

Also fixed up some comments.j

ihc committed:
Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer

Conflicts:
	server/build.gradle

ihc committed:
This is a mix of various things (see below)

Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer (ie. adding svg handling)
Integrated part sourcer with zernike maps.
Fixed javadoc comments that caused errors/added missing javadoc comments.
Changed DEFULT_ZERNIKE_MAP to DEFAULT_ZERNIKE_MAP.
Changed lists to set in Part Sourcer for bulk saving to the db.
Added File utility class for various file related functionality.
Added tests for file utility.
Added constructor tests for better intellij test coverage.
That's all I can remember off the top of my head.
(note that I cannot build it currently due to opencv, so may or may not build ...)
(I DID check it works for the parts that are not dependent on opencv though)

Conflicts:
	server/src/main/java/uk/ac/cam/cl/echo/extrusionfinder/server/parts/Part.java

ihc committed:
Merge branch 'master' of https://github.com/oscarkey/extrusionfinder into sourcer

ihc committed:
Integrated with new grayscaleimagedata loader

ihc committed:
Removed prints in seagatecrawler

Commit:
fixes

ihc committed:
A bit of cleanup, removed PartSourcerTester (no longer meaningful to have unit tests, only integration test)

ihc committed:
Merge branch 'sourcer' of https://github.com/oscarkey/extrusionfinder into sourcer

ihc committed:
Change default crawl max page config back to -1

ihc committed:
And now it compiles as well
